[BOLT] Skip instruction shortening

hilldani opened the pull request:
Add the ability to disable the instruction shortening pass through --shorten-instructions=false

hilldani committed:
add disable pass flags

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                        cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool>
+    DisableFunctionReorder("disable-function-reorder-pass",
```
Function reordering won't run if you don't specify --reorder-functions=<something>

hilldani replied:
function reordering appears to still run, just with the value of "none" which appears to still do something and result in different binary

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                                   cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool>
+    DisableUnreachable("disable-unreachable-pass",
```
--eliminate-unreachable=false accomplishes the same thing

hilldani replied:
removed

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -416,8 +470,10 @@ Error BinaryFunctionPassManager::runAllPasses(BinaryContext &BC) {
-                       opts::EliminateUnreachable);
+  if (!opts::DisableUnreachable) {
+    Manager.registerPass(std::make_unique<EliminateUnreachableBlocks>(PrintUCE),
+                         opts::EliminateUnreachable);
```
When we pass "opts::EliminateUnreachable" here, we're telling the pass manager that this should only run if opts::EliminateUnreachable is true

hilldani replied:
thanks!

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                                    cl::desc("Disable the long jump pass"),
+                                    cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool> DisableBranchFix("disable-branch-fix-pass",
```
This is a mandatory pass

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
                                         cl::ReallyHidden,
                                         cl::cat(BoltOptCategory));
 
+static cl::opt<bool> DisableLongJmp("disable-longjmp-pass",
```
This is a mandatory pass

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                                   cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool>
+    DisableFrameAlloc("disable-alloc-frame-pass",
```
Frame pass is only enabled if you specify --frame-opt

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                              cl::desc("Disable the shorten instruction pass"),
+                              cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool> DisableMemcpy("disable-memcpy-pass",
```
Memcpy pass is only enabled if you pass --inline-memcpy / --memcpy1-spec

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                                      cl::ReallyHidden,
+                                      cl::cat(BoltOptCategory));
+
+static cl::opt<bool> DisableAligner("disable-aligner-pass",
```
This is a mandatory pass

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                                    cl::desc("Disable the aligner pass"),
+                                    cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool> DisableVeneer("disable-veneer-pass",
```
This is a mandatory pass

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                       cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool>
+    DisableAnnotations("disable-annotations-pass",
```
This is a mandatory pass

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                       cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool>
+    DisableNormalizeCFG("disable-normalize-cfg-pass",
```
This is a mandatory pass

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -257,6 +257,56 @@ static cl::opt<bool> CMOVConversionFlag("cmov-conversion",
+                           cl::ReallyHidden, cl::cat(BoltOptCategory));
+
+static cl::opt<bool>
+    DisableShortenInstruction("disable-shorten-instruction-pass",
```
I guess this pass is not mandatory, but it is always enabled because it is usually always a net win. If you want to create a flag to disable it, I suggest creating a flag to enable it "ShortenInstructions" that is default-true, and pass that to the binary pass manager like other passes. Then you can pass --shorten-instructions=false if you want to disable it.

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -53,10 +53,10 @@ extern cl::opt<bool> DumpDotAll;
-DynoStatsAll("dyno-stats-all",
-  cl::desc("print dyno stats after each stage"),
-  cl::ZeroOrMore, cl::Hidden, cl::cat(BoltCategory));
+static cl::opt<bool> DynoStatsAll("dyno-stats-all",
```
Can you limit clang-formatting to your changes only?

rafaelauler reviewed (changes requested):
Hi @hilldani, thanks a lot for submitting your contribution! I've left some comments for you, let me know what you think.

hilldani committed:
closer to normal formatting with some flags refactored

hilldani commented:
Thank you for the review!
1. Removed flags on ones you said were mandatory
2. Refactored remaining flags to be the second argument of Manager.registerPass(...)

hilldani committed:
cleanup

rafaelauler commented:
Failure to build:
https://buildkite.com/llvm-project/github-pull-requests/builds/66438#018fa230-dbc3-443d-8047-627b21ebe471

hilldani committed:
bug fix make function reorder non actually do nothing

hilldani commented:
sorry about that. I tested the compile this time

rafaelauler commented on bolt/lib/Passes/ReorderFunctions.cpp:
```diff
@@ -269,9 +266,11 @@ Error ReorderFunctions::readFunctionOrderFile(
 }
 
 Error ReorderFunctions::runOnFunctions(BinaryContext &BC) {
+  if (opts::ReorderFunctions == RT_NONE) {
```
I tried your change and it looks like there's quite a few tests failing because they depend on the behavior that is currently encoded (to run this logic even if we have RT_NONE). Perhaps this change is not as trivial and it seems we are overlooking some necessary steps here when we have RT_NONE.

If you want to check, you can run ninja check-bolt to see the tests that fail. I suggest we leave this as is and classify reorder functions as mandatory, even when no order is specified.

hilldani replied:
Interesting. Thanks for all the pointers. For the sake of simplicity I'll make this PR just add the option to skip instruction shortening.

hilldani committed:
remove function reorder modifications

rafaelauler commented on bolt/lib/Rewrite/BinaryPassManager.cpp:
```diff
@@ -370,7 +375,7 @@ Error BinaryFunctionPassManager::runAllPasses(BinaryContext &BC) {
     Manager.registerPass(std::make_unique<HugePage>(NeverPrint));
 
-  Manager.registerPass(std::make_unique<ShortenInstructions>(NeverPrint));
+  Manager.registerPass(std::make_unique<ShortenInstructions>(NeverPrint),opts::ShortenInstructions);
```
Use clang-format on this line to get correct formatting (it's missing a space after comma). Remember not to format the entire file, just your changes.

hilldani replied:
formatted

rafaelauler reviewed (approved):
Thanks Daniel! Changes look good to me, just see one nit regarding formatting.

hilldani committed:
clang format

maksfb commented:
Thanks! Please update the summary before committing.

hilldani commented:
update the summary how?

hilldani committed:
Merge branch 'main' into main

maksfb commented:
> update the summary how?

The new description looks good to me. Thanks!

hilldani commented:
not sure how to merge this as I do not have commit access
